src/Crypto/Spake2/Groups/Ed25519.hs

```diff
@@ -62,36 +62,41 @@ ideally demonstrating its equivalence with some automated tests.
 data Ed25519 = Ed25519 deriving (Eq, Show)
 
 instance Group Ed25519 where
-  type Scalar Ed25519 = Integer
   type Element Ed25519 = ExtendedPoint 'Member
 
   elementAdd _ x y = addExtendedPoints x y
   elementNegate group = scalarMultiply group (l - 1)
   groupIdentity _ = assertInGroup extendedZero
-  scalarMultiply _ n x = safeScalarMultiply n x
-
-  integerToScalar _ x = x
-  scalarToInteger _ x = x
 
   encodeElement _ x = encodeAffinePoint (extendedToAffine' x)
   decodeElement _ bytes = toCryptoFailable $ do
     extended <- affineToExtended <$> decodeAffinePoint bytes
     ensureInGroup extended
 
-  generateElement group = do
-    scalar <- generateMax l
-    let element = scalarMultiply group scalar generator
-    pure (KeyPair element scalar)
-
   elementSizeBits _ = 255
-  scalarSizeBits _ = 255
 
   arbitraryElement group bytes =
     let seed = expandArbitraryElementSeed bytes (scalarSizeBytes group + 16) :: ByteString
         y = bytesToNumber seed `mod` q
     in
     List.head [ element | Right element <- map makeGroupMember [y..] ]
 
+instance AbelianGroup Ed25519 where
+  type Scalar Ed25519 = Integer
+
+  scalarMultiply _ n x = safeScalarMultiply n x
+
+  integerToScalar _ x = x
+  scalarToInteger _ x = x
+
+  scalarSizeBits _ = 255
```

Comment:
Yes. I jotted this down after reading something somewhere and forgot to note the source. Oops.

I'm still not 100% sure about Ed25519, but the scalar is *either* an integer between 0 and `q` (2 ^ 255 - 19), *or* an integer between 0 and `l` (2 ^ 252 + 27742317777372353535851937790883648493). 

Either way, it's less than 255 bits. 

Hopefully I'll get around to writing a blog post about Ed25519 and that will help me understand what's going on better.